### DISCUSSION TITLE
How to display HLS m3u8 video in ui.video?

### DISCUSSION THREAD

**Discussion by mrtpk** [zauberzeug/nicegui#3156] [Q&A] [answered]
### Question

Hi Team,

Thank you for sharing your work and I love using nicegui to build UI. I'm stuck with a problem for a few days and I could not find an answer that solves this issue. I appreciate your help. Thanks.

### Context
I'm building a video call application where the user speaks by pressing a button and an output video (frames and audio) is streamed back. I'm streaming the video using ffmpeg to an nginx server. The nginx server converts the stream into HLS format and is provided at "http://localhost:9001/hls/stream.m3u8". I verified the endpoint streaming through VLC and Chrome (version 125.0.6422.114).

### The problem
Following is the code that I tried that uses `ui.video` and `AudioRecorder` is from nicegui sample example:
```py
from nicegui import ui
from nicegui.element import Element
from audio_recorder import AudioRecorder

video = ui.video(src="http://localhost:9001/hls/stream.m3u8", autoplay=True, controls=True) # Doesn't work
# video.props(type="application/x-mpegURL") # raises error

with ui.row().classes('w-full justify-center'):
  audio_recorder = AudioRecorder(on_audio_ready=lambda data: ui.notify(f'Recorded {len(data)} bytes'))
ui.run(dark=False, title='Talker', favicon='🚀', host='0.0.0.0', port=6655, reload=True)
```

No error is displayed in the log or Chrome console. The video element freezes and no controls work.
After some digging I came up with this code:
```py
from nicegui import ui
from nicegui.element import Element
from audio_recorder import AudioRecorder

ui.add_head_html('''
<link href="https://vjs.zencdn.net/8.10.0/video-js.css" rel="stylesheet" />
<style>
.player-container {
  display: flex;
  justify-content: center;
}
</style>                 
''')

video_player_html = '''
  <div class="player-container">
  <video
    id="player"
    class="video-js vjs-default-skin"
    controls
    autoplay
    width="640"
    height="320">

    <source src="http://localhost:9001/hls/stream.m3u8" type="application/x-mpegURL" />

    <p class="vjs-no-js">Cannot load video JS</p>
  </video>
  </div>
'''

script_html = '''
  <script src="https://vjs.zencdn.net/8.10.0/video.min.js"></script>
  <script> 
  var player = videojs('#player');

  player.ready(function() {
  var promise = player.play();

  if (promise !== undefined) {
    promise.then(function() {
      // Autoplay started!
    }).catch(function(error) {
      // Autoplay was prevented.
    });
  }
    });
  </script>
'''

ui.add_body_html(video_player_html) # HTML directly works
ui.add_body_html(script_html)

with ui.row().classes('w-full justify-center'):
  audio_recorder = AudioRecorder(on_audio_ready=lambda data: ui.notify(f'Recorded {len(data)} bytes')) # HTML and audio recording doesn't go well

ui.run(dark=False, title='Talker', favicon='🚀', host='0.0.0.0', port=6655, reload=True)

```

The above code streams correctly but the application is not stable. The "connection lost" dialog pops up when I press the audio record button.  My guess is video.js conflicts with nicegui.

### Question:

What is the best way to use `ui.video` to stream `http://localhost:9001/hls/stream.m3u8`?

I appreciate any help.

**Comment by rodja** [comment]
Thats a tricky question because the setup alone is quite an undertaking. Can you, please, try to provide a self-containing [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)? Ideally without the need to setup a stream ourself. Maybe there is a demo stream available online which you could use to test/demo it?

**Reply by mrtpk** [reply]
Hi @rodja,

Thank you for suggesting to use a demo stream.

The demo stream I tested out is this "https://cph-p2p-msl.akamaized.net/hls/live/2000341/test/master.m3u8". The same behaviour is seen with this URL too. Works in VLC, direct HTML but not through `ui.video`. The modified code:

```py
from nicegui import ui
from nicegui.element import Element
from audio_recorder import AudioRecorder

ui.add_head_html('''
<link href="https://vjs.zencdn.net/8.10.0/video-js.css" rel="stylesheet" />
<style>
.player-container {
  display: flex;
  justify-content: center;
}
</style>                 
''')

video_player_html = '''
  <div class="player-container">
  <video
    id="player"
    class="video-js vjs-default-skin"
    controls
    autoplay
    width="640"
    height="320">

    <source src="https://cph-p2p-msl.akamaized.net/hls/live/2000341/test/master.m3u8" type="application/x-mpegURL" />

    <p class="vjs-no-js">Cannot load video JS</p>
  </video>
  </div>
'''

script_html = '''
  <script src="https://vjs.zencdn.net/8.10.0/video.min.js"></script>
  <script> 
  var player = videojs('#player');

  player.ready(function() {
  var promise = player.play();

  if (promise !== undefined) {
    promise.then(function() {
      // Autoplay started!
    }).catch(function(error) {
      // Autoplay was prevented.
    });
  }
    });
  </script>
'''

# video = ui.video(src="https://cph-p2p-msl.akamaized.net/hls/live/2000341/test/master.m3u8", autoplay=True, controls=True) # Doesn't work

ui.add_body_html(video_player_html) # HTML directly works
ui.add_body_html(script_html)

with ui.row().classes('w-full justify-center'):
  audio_recorder = AudioRecorder(on_audio_ready=lambda data: ui.notify(f'Recorded {len(data)} bytes')) # HTML and audio recording doesn't go well

ui.run(dark=False, title='Talker', favicon='🚀', host='0.0.0.0', port=6655, reload=True)
```

**Comment by falkoschindler** [comment]
Regarding `ui.video`, I think HTML's `<video>` tag doesn't fully support M3U8 streams with all browsers. This works in Safari, but not in Chrome:
```html
<html>
  <body>
    <video width="320" height="240" controls>
      <source src="https://cph-p2p-msl.akamaized.net/hls/live/2000341/test/master.m3u8" />
      Your browser does not support the video tag.
    </video>
  </body>
</html>
```
This might be the reason for `ui.video` not showing such streams.

Your workaround using video.js seems to work. I can't reproduce the connection problem you're describing. But I don't know this library. Maybe there's some more magic going on causing problems under certain conditions.

**Reply by mrtpk** [reply]
Thank you @falkoschindler, you are right. It did work on Safari but not on Chrome or Edge.

**Comment by mrtpk** [comment, marked as answer]
From the discussion, I understood that `ui.video`, underneath uses HTML 5 video tag. The video tag is implemented differently in different browsers. The m3u8 HLS stream is not supported by HTML5 video tag in Chrome or Edge yet; hence, it was not streaming. But the tag is supported in Safari, so `ui.video` streams m3u8 on Safari.

Thank you @rodja and @falkoschindler for your help. Thanks.
